zipkin/src/main/java/zipkin/storage/InMemorySpanStore.java

```diff
@@ -38,16 +41,45 @@
 import static zipkin.internal.GroupByTraceId.TRACE_DESCENDING;
 import static zipkin.internal.Util.sortedList;
 
-/** Internally, spans are indexed on 64-bit trace ID */
+/**
+ * Internally, spans are indexed on 64-bit trace ID
+ *
+ * Here's an example of some traces in memory:
+ *
+ * traceIdToSpans:
+ *    aaaa --> [ span1(time:July 4, traceId:aaaa, service:foo, name:GET),
+ *               span2(time:July 5, traceId:aaaa, service:bar, name:GET) ]
+ *    bbbb --> [ span3(time:July 5, traceId:bbbb, service:biz, name:GET),
+ *               span4(time:July 6, traceId:bbbb) service:foo, name:POST]
+ *    cccc --> [ span5(time:July 4, traceId:aaaa, service:foo, name:GET)]
+ *
+ * traceIdTimeStamps: [ (aaaa, July 4), (aaaa, July 5), (bbbb, July 5),
+ *                      (bbbb, July 6), (cccc, July 4) ]
+ *
+ * serviceToTraceIdTimeStamp:
+ *    foo --> [ (aaaa, July 4), (cccc, July 4), (bbbb, July 6) ]
+ *    bar --> [ (aaaa, July 5) ]
+ *    biz --> [ (bbbb, July 5) ]
+ *
+ * serviceToSpanNames:
+ *    foo --> [ GET, POST ]
+ *    bar --> [ GET ]
+ *    biz --> [ GET ]
+ */
 public final class InMemorySpanStore implements SpanStore {
   private final Multimap<Long, Span> traceIdToSpans = new LinkedListMultimap<>();
-  private final Set<Pair<Long>> traceIdTimeStamps = new TreeSet<>(VALUE_2_DESCENDING);
-  private final Multimap<String, Pair<Long>> serviceToTraceIdTimeStamp =
+  private final TreeSet<Pair<Long>> traceIdTimeStamps = new TreeSet<>(VALUE_2_DESCENDING);
+  private final SortedByValue2Descending serviceToTraceIdTimeStamp =
       new SortedByValue2Descending<>();
   private final Multimap<String, String> serviceToSpanNames =
       new LinkedHashSetMultimap<>();
   private final boolean strictTraceId;
+  private final StorageMetrics metrics;
+  private final int maxSpans;
   volatile int acceptedSpanCount;
+  volatile int retainedSpanCount;
```

Comment:
this feels like a fair amount of bookkeeping which may make this class even less attractive to maintain. See how much of this you can get rid of and have it still work